Implement backend fallback fallthrough

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32439 Implement backend fallback fallthrough**

This patch makes it possible to register a fallthrough kernel as a backend fallback, so that if we don't define a specific operator definition for that kernel, we just skip the backend entirely.  This is implemented in a such a way that it costs no more to fallthrough than to go straight to the actual implementation kernel; this is important for our intended use case, CompoundOp, as this backend will be enabled on ALL operations but only bit hit in a minority of cases. Design at https://github.com/pytorch/pytorch/issues/29548

The implementation may seem a little complicated. Here is the general strategy:

* A "fallthrough kernel" is a kernel which redispatches to the next eligible dispatch key.  This is a useful alternative to the current default behavior, which is to error when a kernel is not found. For example, we want to register compound ops in the dispatcher, and have them be processed preferentially before any other dispatch keys. Since not all ops have compound implementations, we need to fallthrough in those cases (not error!)
* We could implement fallthrough kernels as a boxed kernel that redispatches. However, this is a bit inefficient, since we will do a lot of repeated boxing and unboxing in that case. A better strategy is to mask out the keys in the dispatch key set that correspond to fallthrough kernels, so we simply do not consider them at all when deciding where to go.
* How do we determine this mask? Here things get a bit complicated. At the highest level, we want to delete a dispatch key from consideration ONLY if we have registered a fallthrough kernel for it, and not otherwise. But if we register a backend-wide fallthrough kernel, and then specifically override one or two operations on it to have custom implementations, we better not delete the key. So we have to do the mask in two phases: we first have a mask like 11111011, with zeros wherever a fallthrough kernel has been registered backend-wide, and then another mask like 00000100 (or perhaps 00000000) with a one wherever we have registered a op-specific kernel, and OR the masks together; the latter mask "undos" the zero in the fallthrough mask.
* The strategy above is incomplete; it will not skip dispatch if you register a fallthrough kernel for a specific operator. But we have no known use case for this operation, and https://github.com/pytorch/pytorch/issues/32454 should make things better.

There's a little bit of abstraction confusion, since DispatchTable has to ensure that DispatchKeyExtractor has an up-to-date understanding of what per-operator kernel operations have been registered. I also didn't really update the "public" API to permit this, so you still have to use the internal API to program against this.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19503886](https://our.internmc.facebook.com/intern/diff/D19503886)

ezyang committed:
Implement fallthrough kernel.

This adds c10::fallthrough_kernel which is a special boxed function which
can be used to implement fallthrough behavior at a dispatch key.  A fallthrough
kernel will redispatch to the next valid dispatch key.  It is implemented
in such a way that it costs no more to fallthrough than it does to go
straight to the actual implementation of the kernel.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

izdeby commented:
Is there any way we can add tests to this? Maybe a test plan in the description?

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit ab5d5693:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ab5d5693759e0b4ce2beb3cfda7b08f933d1e4b5).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4305016) pytorch_linux_backward_compatibility_check_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=88911353) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=88911353))
<details>
<summary>
<code>Jan 21 17:45:44 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.</code>
</summary>

```
Jan 21 17:45:44 processing existing schema:  aten::values(Dict(str, t) self) -> (t[](*)) 
Jan 21 17:45:44 processing existing schema:  aten::values(Dict(int, t) self) -> (t[](*)) 
Jan 21 17:45:44 processing existing schema:  aten::values(Dict(float, t) self) -> (t[](*)) 
Jan 21 17:45:44 processing existing schema:  aten::values(Dict(Tensor, t) self) -> (t[](*)) 
Jan 21 17:45:44 processing existing schema:  aten::view(Tensor(a) self, int[] size) -> (Tensor(a)) 
Jan 21 17:45:44 processing existing schema:  aten::where(Tensor condition) -> (Tensor[]) 
Jan 21 17:45:44 processing existing schema:  aten::where.self(Tensor condition, Tensor self, Tensor other) -> (Tensor) 
Jan 21 17:45:44 processing existing schema:  aten::zeros_like(Tensor self, *, int? memory_format=None) -> (Tensor) 
Jan 21 17:45:44 processing existing schema:  aten::zeros_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=None) -> (Tensor) 
Jan 21 17:45:44 processing existing schema:  prim::Constant() -> (...) 
Jan 21 17:45:44 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.  
Jan 21 17:45:44  
Jan 21 17:45:44 Broken ops: [ 
Jan 21 17:45:44 	_TorchScriptTesting_StackString::merge(torch.classes._TorchScriptTesting_StackString _0, torch.classes._TorchScriptTesting_StackString _1) -> () 
Jan 21 17:45:44 	_TorchScriptTesting_StackString::clone(torch.classes._TorchScriptTesting_StackString _0) -> (torch.classes._TorchScriptTesting_StackString _0) 
Jan 21 17:45:44 ] 
Jan 21 17:45:44 + cleanup 
Jan 21 17:45:44 + retcode=1 
Jan 21 17:45:44 + set +x 
Jan 21 17:45:44 =================== sccache compilation log =================== 
Jan 21 17:45:44 =========== If your build fails, please take a look at the log above for possible reasons =========== 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

ezyang commented:
Working on tests right now :)

ezyang committed:
Update on "Implement fallthrough kernel."


This adds c10::fallthrough_kernel which is a special boxed function which
can be used to implement fallthrough behavior at a dispatch key.  A fallthrough
kernel will redispatch to the next valid dispatch key.  It is implemented
in such a way that it costs no more to fallthrough than it does to go
straight to the actual implementation of the kernel.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
Tests added

ezyang commented:
Actually, for unblocking this implements the wrong functionality; what izdeby and mcarilli is backend fallback fallthrough, not per-operator fallthrough. I can only conveniently support one of these. I should support backend fallback fallthrough.

ezyang committed:
Update on "Implement fallthrough kernel."


This adds c10::fallthrough_kernel which is a special boxed function which
can be used to implement fallthrough behavior at a dispatch key.  A fallthrough
kernel will redispatch to the next valid dispatch key.  It is implemented
in such a way that it costs no more to fallthrough than it does to go
straight to the actual implementation of the kernel.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19503886](https://our.internmc.facebook.com/intern/diff/D19503886)

[ghstack-poisoned]

ezyang committed:
Update on "Implement backend fallback fallthrough"


This patch makes it possible to register a fallthrough kernel as a backend fallback, so that if we don't define a specific operator definition for that kernel, we just skip the backend entirely.  This is implemented in a such a way that it costs no more to fallthrough than to go straight to the actual implementation kernel; this is important for our intended use case, CompoundOp, as this backend will be enabled on ALL operations but only bit hit in a minority of cases.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19503886](https://our.internmc.facebook.com/intern/diff/D19503886)

[ghstack-poisoned]

ezyang committed:
Update on "Implement backend fallback fallthrough"


This patch makes it possible to register a fallthrough kernel as a backend fallback, so that if we don't define a specific operator definition for that kernel, we just skip the backend entirely.  This is implemented in a such a way that it costs no more to fallthrough than to go straight to the actual implementation kernel; this is important for our intended use case, CompoundOp, as this backend will be enabled on ALL operations but only bit hit in a minority of cases.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19503886](https://our.internmc.facebook.com/intern/diff/D19503886)

[ghstack-poisoned]

ezyang commented:
Correct version now implemented

ezyang committed:
Update on "Implement backend fallback fallthrough"


This patch makes it possible to register a fallthrough kernel as a backend fallback, so that if we don't define a specific operator definition for that kernel, we just skip the backend entirely.  This is implemented in a such a way that it costs no more to fallthrough than to go straight to the actual implementation kernel; this is important for our intended use case, CompoundOp, as this backend will be enabled on ALL operations but only bit hit in a minority of cases. Design at https://github.com/pytorch/pytorch/issues/29548

The implementation may seem a little complicated. Here is the general strategy:

* A "fallthrough kernel" is a kernel which redispatches to the next eligible dispatch key.  This is a useful alternative to the current default behavior, which is to error when a kernel is not found. For example, we want to register compound ops in the dispatcher, and have them be processed preferentially before any other dispatch keys. Since not all ops have compound implementations, we need to fallthrough in those cases (not error!)
* We could implement fallthrough kernels as a boxed kernel that redispatches. However, this is a bit inefficient, since we will do a lot of repeated boxing and unboxing in that case. A better strategy is to mask out the keys in the dispatch key set that correspond to fallthrough kernels, so we simply do not consider them at all when deciding where to go.
* How do we determine this mask? Here things get a bit complicated. At the highest level, we want to delete a dispatch key from consideration ONLY if we have registered a fallthrough kernel for it, and not otherwise. But if we register a backend-wide fallthrough kernel, and then specifically override one or two operations on it to have custom implementations, we better not delete the key. So we have to do the mask in two phases: we first have a mask like 11111011, with zeros wherever a fallthrough kernel has been registered backend-wide, and then another mask like 00000100 (or perhaps 00000000) with a one wherever we have registered a op-specific kernel, and OR the masks together; the latter mask "undos" the zero in the fallthrough mask.
* The strategy above is incomplete; it will not skip dispatch if you register a fallthrough kernel for a specific operator. But we have no known use case for this operation, and https://github.com/pytorch/pytorch/issues/32454 should make things better.

There's a little bit of abstraction confusion, since DispatchTable has to ensure that DispatchKeyExtractor has an up-to-date understanding of what per-operator kernel operations have been registered. I also didn't really update the "public" API to permit this, so you still have to use the internal API to program against this.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19503886](https://our.internmc.facebook.com/intern/diff/D19503886)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/boxing/KernelFunction.cpp:
```diff
@@ -0,0 +1,18 @@
+void fallthrough_kernel(OperatorKernel*, const OperatorHandle&, Stack*) {
+  TORCH_INTERNAL_ASSERT(0,
+    "fallthrough_kernel was executed but it should have been short-circuited by the dispatcher. "
+    "This could occur if you registered a fallthrough kernel as a backend fallback; this is NOT "
```
wait, registering it as backend fallback is the main purpose. You probably mean "This could occur if you registered a fallthrough kernel as a backend kernel for an operator".

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.cpp:
```diff
@@ -0,0 +1,13 @@
+
+namespace c10 {
+
+void DispatchKeyExtractor::setIsOperatorOverridden(DispatchKey k, bool is_overridden) {
```
I find the name a bit confusing since this is not strictly about overriding backend fallbacks, this is also set if a backend doesn't even have a fallback but the operator has a kernel for this key. Maybe a better name is `setOperatorHasKernelForBackend` (and also rename the member variable).

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -21,9 +22,18 @@ namespace impl {
+    // These exclusions are NOT tracked in the TLS, but must be applied AFTER
+    // TLS, which is why you have to pass them in to this function (as opposed
+    // to just applying it to the input 'ts').
+    DispatchKeySet non_fallthrough_mask
```
from a naming point of view, I wouldn't name this a mask. You're using it as a mask but that is an implementation detail that callers don't know about. The argument (from a callers perspective) is semantically a set of dispatch keys, namely the dispatch keys that don't have a fallthrough kernel registered. So maybe call it `backends_without_fallthrough` or something like that? That would also allow you to avoid or shorten the comment explaining that the mask is inverted.

smessmer commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -84,33 +94,52 @@ struct DispatchKeyExtractor final {
+  // per-op basis, but while we could fix this by maintaining a second
+  // DispatchKeySet, it doesn't seem that there is any actual use case,
+  // so we are deferring it for 32454.
+  DispatchKeySet perOperatorOverriddenKernels_;
```
I find the name a bit confusing since this is not strictly about overriding backend fallbacks, this is also set if a backend doesn't even have a fallback but the operator has a kernel for this key. Maybe a better name is `setOperatorHasKernelForBackend` and `operatorHasKernelForBackends_`.

After that naming change, the comment doesn't need to be at the member variable anymore because the member variable is self-explanatory, so the comment can be moved to where the variable is used, i.e. into dispatchKeySetToDispatchKey_. That's probably where a reader of the code would wonder why you're or-ing this mask.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -144,6 +144,8 @@ class CAFFE2_API Dispatcher final {
   std::list<OperatorDef> operators_;
   LeftRight<ska::flat_hash_map<OperatorName, OperatorHandle>> operatorLookupTable_;
   impl::KernelFunctionTable backendFallbackKernels_;
+  // Set of backends which have specified they want fallthrough behavior
```
I think it's the set of backends that **don't** have specified they want fallthrough behavior.

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -32,6 +32,7 @@ Dispatcher::Dispatcher()
 : operators_()
 , operatorLookupTable_()
 , backendFallbackKernels_()
+, backendFallbackNonFallthroughSet_(DispatchKeySet::FULL)
```
This name and the `backend_nonfallthrough_mask` name above could probably be unified into one name to make clear that those are the same thing. I like this name here better than the `backend_nonfallthrough_mask` above but something like `backendsWithoutFallthrough` is probably even shorter and clearer.

smessmer reviewed (commented):
Implementation looks very good, just some naming suggestions.

ezyang commented:
I regretfully most conclude that this PR is likely the cause of 

```
Jan 22 08:30:42 ======================================================================
Jan 22 08:30:42 ERROR [0.015s]: test_cos_scalar_cpu (__main__.TestAutogradDeviceTypeCPU)
Jan 22 08:30:42 ----------------------------------------------------------------------
Jan 22 08:30:42 Traceback (most recent call last):
Jan 22 08:30:42   File "/Users/distiller/project/test/common_device_type.py", line 180, in instantiated_test
Jan 22 08:30:42     return test(self, device_arg)
Jan 22 08:30:42   File "test_autograd.py", line 4011, in do_test
Jan 22 08:30:42     check(name)
Jan 22 08:30:42   File "test_autograd.py", line 3936, in check
Jan 22 08:30:42     output_variable, (self_variable,) + args_variable)
Jan 22 08:30:42   File "test_autograd.py", line 3866, in run_grad_and_gradgrad_checks
Jan 22 08:30:42     test_case.assertTrue(gradgradcheck(apply_method, input_variables, gen_non_contig_grad_outputs=True))
Jan 22 08:30:42   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 407, in gradgradcheck
Jan 22 08:30:42     nondet_tol=nondet_tol)
Jan 22 08:30:42   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 297, in gradcheck
Jan 22 08:30:42     'The tolerance for nondeterminism was {}.'.format(nondet_tol))
Jan 22 08:30:42   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 229, in fail_test
Jan 22 08:30:42     raise RuntimeError(msg)
Jan 22 08:30:42 RuntimeError: Backward is not reentrant, i.e., running backward with same input and grad_output multiple times gives different values, although analytical gradient matches numerical gradient. The tolerance for nondeterminism was 0.0.
Jan 22 08:30:42 
Jan 22 08:30:42 ----------------------------------------------------------------------
```

https://github.com/pytorch/pytorch/issues/32502
